// eslint-plugin-lodash/lib/rules/import-scope.js
/**
 * @fileoverview Rule to disallow the use of a chain for a single method
 */
'use strict';

/**
 * @fileoverview Rule to disallow the use of a chain for a single method
 */
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

var _require = require('../util/importUtil'),
    isFullLodashImport = _require.isFullLodashImport,
    getNameFromCjsRequire = _require.getNameFromCjsRequire,
    getMethodImportFromName = _require.getMethodImportFromName;

var every = require('lodash/every');
var includes = require('lodash/includes');

var messages = {
    method: 'Do not import from the full Lodash module.',
    member: 'Import members from the full Lodash module.',
    full: 'Use the full Lodash module.'
};

var importNodeTypes = {
    method: ['ImportDefaultSpecifier'],
    member: ['ImportSpecifier'],
    full: ['ImportDefaultSpecifier', 'ImportNamespaceSpecifier']
};

var isMethodImport = function isMethodImport(name) {
    return Boolean(getMethodImportFromName(name));
};
var allImportsAreOfType = function allImportsAreOfType(node, types) {
    return every(node.specifiers, function (specifier) {
        return includes(types, specifier.type);
    });
};

module.exports = {
    meta: {
        schema: [{
            enum: ['method', 'member', 'full']
        }]
    },
    create: function create(context) {
        var importType = context.options[0] || 'method';

        return {
            ImportDeclaration: function ImportDeclaration(node) {
                if (isFullLodashImport(node.source.value)) {
                    if (importType === 'method') {
                        context.report(node, messages.method);
                    } else {
                        if (!allImportsAreOfType(node, importNodeTypes[importType])) {
                            context.report(node, messages[importType]);
                        }
                    }
                } else if (isMethodImport(node.source.value) && importType !== 'method') {
                    context.report(node, messages[importType]);
                }
            },
            VariableDeclarator: function VariableDeclarator(node) {
                var name = getNameFromCjsRequire(node.init);
                if (isFullLodashImport(name)) {
                    if (importType === 'method') {
                        context.report(node, messages.method);
                    } else {
                        var isObjectPattern = node.id.type === 'ObjectPattern';
                        var isMemberImport = importType === 'member';
                        if (isObjectPattern !== isMemberImport) {
                            context.report(node, messages[importType]);
                        }
                    }
                } else if (isMethodImport(name) && importType !== 'method') {
                    context.report(node, messages[importType]);
                }
            }
        };
    }
};
